UI component being attempted to be disposed in a non-ui thread.

deepika-u opened the pull request:
Fixes #2316

As suggested tried fixing in destroy(), but looks like we need extra testing to ensure no shutdown regressions considering the below
- In shutdown, syncExec might block if the display is already disposed or unresponsive.
- EventBroker cleanup still runs in non-UI thread, which is fine. But we have 2 different parts happening - part disposed later, eventbroker unsubscribed immediately.

Note : I have no recreate to test this fix, can this be tested with the existing Unit Test infrastructure in place already?

deepika-u committed:
UI component being attempted to be disposed in a non-ui thread.

deepika-u commented:
@jukzi : can you take a look at this when you have some time please?

merks commented:
Note that he is "missing in action" so don't expect anything from him.

laeubi commented on bundles/org.eclipse.ui.workbench/eclipseui/org/eclipse/ui/internal/e4/compatibility/CompatibilityPart.java:
```diff
@@ -420,7 +421,12 @@ public void create() {
+				display.syncExec(this::invalidate);
+			} else {
+				invalidate();
+			}
```
E4 Has the `UISynchronize` service for this purposes:


```suggestion
void destroy(UISynchronize sync) {
if (!alreadyDisposed) {
sync.syncExec ...
```

I also wonder if not the caller is to blame not calling from the UI thread, so maybe this should better be fixed than on the client side.

deepika-u replied:
yes, thanks for the newer way. updated.

jukzi commented on bundles/org.eclipse.ui.workbench/eclipseui/org/eclipse/ui/internal/e4/compatibility/CompatibilityPart.java:
```diff
@@ -420,7 +421,12 @@ public void create() {
 		if (!alreadyDisposed) {
-			invalidate();
+			Display display = Display.getDefault();
+			if (display != null && !display.isDisposed() && Display.getCurrent() != display) {
```
typically the isDisposed() check has to be repeated in Display thread, because dispose could happen in the diplay thread meanwhile

deepika-u replied:
> typically the isDisposed() check has to be repeated in Display thread, because dispose could happen in the diplay thread meanwhile

added another check as well.

deepika-u commented:
@laeubi @jukzi : all the comments are incorporated, can you check now please. Thanks for your suggestions.

laeubi commented:
@deepika-u I still think it is better fixed at the caller side than in this particular method, so in what cases can it really happen that it is called from non-ui thread?

deepika-u commented:
> @deepika-u I still think it is better fixed at the caller side than in this particular method, so in what cases can it really happen that it is called from non-ui thread?

CompatibilityPart.destroy()
 → InjectorImpl.disposed()
 → EclipseContext.dispose()
 → OSGI/Equinox bundle stop thread
That path is not on the UI thread. It’s on the OSGi framework thread during bundle shutdown. So yes, it can and does happen that destroy() can be called from non-UI thread. That’s how we got the NPE in the first place.

In general 
- In normal user-driven part closure, destroy() will indeed be called from the UI thread.
- But during application or bundle shutdown, destroy() can be invoked from OSGi/framework threads (non-UI). That’s exactly how the reported NPE occurred.
- Since IntroPart.dispose() and other cleanup touch SWT/JFace resources are all dealing with UI components, they must be marshaled to the UI thread.

Therefore, the correct place to fix is in CompatibilityPart.destroy, which bridges the generic e4 lifecycle and the UI-thread-sensitive 3.x parts.

This way we can ensure 
- Lifecycle events are still respected.
- invalidate() always runs in UI thread.
- No raw Display usage, just e4-provided UISynchronize.

laeubi commented:
> But during application or bundle shutdown, destroy() can be invoked from OSGi/framework threads (non-UI). That’s exactly how the reported NPE occurred.

I think you should find out where exactly this enters the system and what is triggering that part. There should be one place where we can ensure it is called inside the UI thread.

So can you share a stacktrace of the problematic call so we can better decide what would be the best place here?

deepika-u commented:
@laeubi 
> So can you share a stacktrace of the problematic call so we can better decide what would be the best place here?

I tried myself running CTabItemTest  and captured these logs.
I dint find anything related to the error in .log file. But could see the same error in console of eclipse. So just copied the content of console to a file and also sharing the .log file for reference.
[console_output.txt](https://github.com/user-attachments/files/22115442/console_output.txt)
[.log](https://github.com/user-attachments/files/22115449/default.log)

Please have a look when you get some time and let me know your opinion.

deepika-u commented:
@laeubi 
On further anlaysis, i felt EditActionGroup.dispose() also needs to be updated. So adding this change too to make the fix complete.
Now with the EditActionGroup.dispose() changes in place along with CompatibilityPart.destroy(), i am not able to get the original reported NPE anymore on running the same CTabItemTest. Attaching the console for the same -> 
[Console_output_4thSept.txt](https://github.com/user-attachments/files/22138345/Console_output_4thSept.txt)

Once you get some time, can you review the changes please.

But i see another problem here though not very important but still a problem "org.eclipse.swt.SWTException: Widget is disposed"
For addressing this problem i have a pr attached which adds safe guards but no functionality being updated via it.
Will be adding the pr once raised. 

#3241 is created for this.

laeubi commented:
In the log I can only see errors about already disposed part / device but not any access from another thread?

deepika-u commented:
> In the log I can only see errors about already disposed part / device but not any access from another thread?

yes exactly now with these both file changes, i dont see NPE anymore.

laeubi commented:
> > In the log I can only see errors about already disposed part / device but not any access from another thread?
> 
> yes exactly now with these both file changes, i dont see NPE anymore.

The point is that we should not fix side-effects but the callers who call the method outside the Device thread so that callers stacktrace would be required here to decide where to best fix it. Requiring everyone dispose method (and you already discovered now two places) does not really scale well.

deepika-u commented:
@laeubi 
> The point is that we should not fix side-effects but the callers who call the method outside the Device thread so that callers stacktrace would be required here to decide where to best fix it. Requiring everyone dispose method (and you already discovered now two places) does not really scale well.

May be i missed your intent and didnt understand you clearly. But what i also tried as updated in https://github.com/eclipse-platform/eclipse.platform.ui/pull/3232#issuecomment-3248623090

I have tried running /org.eclipse.e4.ui.tests.css.swt/src/org/eclipse/e4/ui/tests/css/swt/CTabItemTest.java locally to reproduce the NPE. As said by jukzi in https://github.com/eclipse-platform/eclipse.platform.ui/issues/2316#issue-2542988909

Please do let me know if i have to try out anything else or capture any other logs to be specific.

laeubi commented:
> I have tried running /org.eclipse.e4.ui.tests.css.swt/src/org/eclipse/e4/ui/tests/css/swt/CTabItemTest.java locally to reproduce the NPE

And was this successful? Sorry to ask but its a bit unclear to me if the problem is currently happens, only happens in the test and how it relates to `EditActionGroup`?

deepika-u commented:
> And was this successful? Sorry to ask but its a bit unclear to me if the problem is currently happens, only happens in the test and how it relates to `EditActionGroup`?

From this comment -> https://github.com/eclipse-platform/eclipse.platform.ui/pull/3232#issuecomment-3248623090
If you go through the full description in this section i have ended up in adding the change in EditActionGroup. Respective log for reference is also attached.

Now with CompatibilityPart and EditActionGroup changes in place then attempt to run /org.eclipse.e4.ui.tests.css.swt/src/org/eclipse/e4/ui/tests/css/swt/CTabItemTest.java locally i am not seeing the NPE as reported in https://github.com/eclipse-platform/eclipse.platform.ui/issues/2316

But i see errors related to "org.eclipse.swt.SWTException: Widget is disposed" which i am trying to address with pr https://github.com/eclipse-platform/eclipse.platform.ui/pull/3241

Never mind, thanks for asking again.

Observation : Now with both the pr changes in place when i run CTabItemTest locally, i dont see both the errors. It is clean.

laeubi commented:
@deepika-u now we have

- https://github.com/eclipse-platform/eclipse.platform.ui/pull/3241

is this PR actually needed?

deepika-u commented:
@laeubi 
Thanks for merging #3241 
> is this PR actually needed?

Yes definitely this would be needed. Because both are addressing 2 different problems.
Today i have updated my master to latest code.
Now when i run CTabItemTest on master, i am still seeing the java.lang.RuntimeException caused by CompatibilityPart and EditActionGroup. Attached the console output here -> 
[console_output_8thSept.txt](https://github.com/user-attachments/files/22201749/console_output_8thSept.txt)

Now i have updated my branch to be in sync with master and now when i run CTabItemTest, i dont see any exceptions as such.
For reference also attached console output -> 
[console_output_8thSept_onbranch_3232.txt](https://github.com/user-attachments/files/22201901/console_output_8thSept_onbranch_3232.txt)

So this confirms that this pr is also needed for successful run of CTabItemTest along with #3241. Hope i have clarified. Thanks for the ask.

laeubi commented:
> Now when i run CTabItemTest on master, i am still seeing the java.lang.RuntimeException caused by CompatibilityPart and EditActionGroup. 

What I'm wondering about is the following:

1. Do we see it anywhere when running Eclipse at all?
2. Why does it not happen in the ibuild but only in the IDE?
3. Should maybe the tests be simply marked to be running in the UI thread instead?

deepika-u commented:
@laeubi 
We don’t see this issue in day-to-day Eclipse usage - it only shows up in JUnit test runs because shutdown timing differs between the IDE and the ibuild environment. The change we introduced makes disposal more robust (guards against disposed widgets and device access), so it fixes the test failures without negative side effects. Marking the test as “UI-thread only” would only mask the problem; our fix ensures correctness in both runtime and test scenarios with no behavioral change.

laeubi commented on bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/EditActionGroup.java:
```diff
@@ -55,11 +57,25 @@ public EditActionGroup(Shell aShell) {
+				}
+			}
 		}
-		super.dispose();
```
Why not call super.dispose() anymore?

deepika-u replied:
@laeubi 
The problem if we still call "super.dispose();" inside dispose(), it would run immediately - before the clipboard is actually disposed, since disposal happens asynchronously later.
This creates a lifecycle ordering ambiguity:
Context cleared too early.
Asynchronous disposal runs after super.dispose()

If that async runnable references anything indirectly tied to the context (logging, resources, actions), you can get into NPEs or disposed-object access.

So this is Why super.dispose() was dropped.
- super.dispose() is trivial (just setContext(null)).
- Avoided calling it before async disposal finishes, to prevent race conditions.
- In practice, leaving context uncleared is not a big resource leak compared to crashing the UI.

So it was sacrificed for safety.

But if we still want to be correct, the better solution is to still call super.dispose() inside the async block. That way we don’t silently skip the context cleanup from ActionGroup.

We can do this way ::
```
@Override
public void dispose() {
    Display display = Display.getDefault();
    if (display != null && !display.isDisposed()) {
        display.asyncExec(() -> {
            if (clipboard != null && !clipboard.isDisposed()) {
                clipboard.dispose();
                clipboard = null;
            }
            EditActionGroup.super.dispose();
        });
    } else {
        // Display already gone, dispose synchronously as a fallback
        if (clipboard != null && !clipboard.isDisposed()) {
            try {
                clipboard.dispose();
                clipboard = null;
            } catch (SWTException e) {
                // Log or ignore safely
            }
        }
        super.dispose();
    }
}
```

laeubi commented:
> our fix ensures correctness in both runtime and test scenarios with no behavioral change.

When disposal happens during call of dispose or async / blocking in UI running alongside could of course seen as behavior change. And if Test + real environments differ there is of course a chance for false positives (or we did not capture problems). Also it might have side-effects and complicates the code.

e.g. in your current change I can see

```
Display display = Display.getDefault();
if (display != null && !display.isDisposed()) {

}
```

if you look at javadoc you will see

> Returns the default display. **One is created (making the thread that invokes this method its user-interface thread) if it did not already exist**.

so your check for `null` will never succeed. If no display exits one will be created here (and maybe never disposed). So its always hard to guess its really "side-effect-free" to add such code.

laeubi commented on bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/EditActionGroup.java:
```diff
@@ -55,11 +57,25 @@ public EditActionGroup(Shell aShell) {
-			clipboard = null;
+		Display display = Display.getDefault();
+		if (display != null && !display.isDisposed()) {
+			display.asyncExec(() -> {
```
In the otherplace `syncExec` is used, why is this different here?

deepika-u replied:
EditActionGroup.dispose() with asyncExec() because safety during shutdown is more important than ordering.
- Disposing the Clipboard (and similar UI resources) must happen on the UI thread.
- But dispose() might be called from a non-UI thread during shutdown.
- If we used syncExec() here and called it from a non-UI thread while the display is already disposing, it could deadlock (the UI thread might already be blocked, tearing down widgets).
- asyncExec() safely queues the disposal for "later" on the UI thread, avoiding blocking and deadlock risk.
- If the display is already gone, we just try/catch to clean up as best we can.
So in summary asyncExec() => fire and forget cleanup, safe during shutdown.

CompatibilityPart.destroy() with syncExec() because ordering and completion are crucial for consistent teardown.
- destroy() is part of the Eclipse 4 lifecycle (@PreDestroy), which means the framework expects the part to be fully disposed when this method returns.
- If you used asyncExec() here, invalidate() would run later, and there’s a risk that other teardown code(unsubscribes, model cleanup, etc.) would happen before invalidate() ran. That could cause inconsistencies or NPEs.
- By using syncExec(), you ensure that invalidate() actually completes before the method continues and the lifecycle is maintained correctly.
So in summary syncExec() => enforce order, synchronous cleanup before continuing.

deepika-u commented:
@laeubi 
> Display display = Display.getDefault();
> if (display != null && !display.isDisposed()) {
> 
> }
> ```
> if you look at javadoc you will see
> 
> > Returns the default display. **One is created (making the thread that invokes this method its user-interface thread) if it did not already exist**.

Thanks for the feedback! You're absolutely right about the side effects of Display.getDefault(). I've updated the code to use Display.getCurrent() instead, which avoids creating a new display and ensures we're only disposing UI resources on the UI thread. This should reduce complexity and avoid unintended behavior in both test and runtime environments.

When you get some time, could you take a look at it now please?

laeubi commented on bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/EditActionGroup.java:
```diff
@@ -55,9 +57,22 @@ public EditActionGroup(Shell aShell) {
-		if (clipboard != null) {
-			clipboard.dispose();
-			clipboard = null;
+		Display display = Display.getCurrent();
```
why not using `shell.getDisplay()` it seems obvious here?

deepika-u replied:
Yeah shell.getDisplay() is also a good choice. But i need to ensure that the Shell reference is not null and not disposed before calling shell.getDisplay(). Can i go ahead with this way?

laeubi replied:
Thinking more can you use clipboard itself, it is already is disposed what is the point doing more work here?

deepika-u replied:
@leubi 
Sorry, was not able to get your intent.  You meant to use Display.getCurrent() or shell.getDisplay()?

deepika-u replied:
If we dont have this pr in place, while running CTabItemTest locally we'll again endup into original problem reshaped as( post pr 3241)  
this problem shown in the console output below -> 
[Console_without_pr3232.txt](https://github.com/user-attachments/files/22402471/Console_without_pr3232.txt)

laeubi replied:
@deepika-u where do you get this is it windows? 
I think it makes sense to go through each exception step by step, the first I see in your log is:

```
org.eclipse.swt.SWTException: Device is disposed
	at org.eclipse.swt.SWT.error(SWT.java:4946)
	at org.eclipse.swt.SWT.error(SWT.java:4861)
	at org.eclipse.swt.SWT.error(SWT.java:4832)
	at org.eclipse.swt.widgets.Display.error(Display.java:1332)
	at org.eclipse.swt.widgets.Display.getThread(Display.java:2696)
	at org.eclipse.swt.dnd.Clipboard.dispose(Clipboard.java:219)
	at org.eclipse.ui.internal.navigator.resources.actions.EditActionGroup.dispose(EditActionGroup.java:59)
```

This looks like a bug in SWT (windows), because calling dispose on an already disposed component should be a no-op (and if device is disposed clipboard is gone anyways...).

Looking into other widgets implementations the do

`if (display.thread != Thread.currentThread ()) error (SWT.ERROR_THREAD_INVALID_ACCESS);`

so calling `getThread` instead of access the field directly seems the culprit here.

laeubi commented on bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/EditActionGroup.java:
```diff
@@ -55,9 +57,22 @@ public EditActionGroup(Shell aShell) {
+			// Non-UI thread or display disposed
+			if (clipboard != null && !clipboard.isDisposed()) {
+				try {
+					clipboard.dispose();
```
Can the clipboard really be disposed outside the UI thread?!?

deepika-u replied:
Ideally not supposed to be which might result into SWTException or resource leaks if disposal fails silently or false positives. 
In my opinion - If we are not on the UI thread and can't guarantee access to it, it's safer to skip disposal or log a warning.

Do you mean i can skip this line `clipboard.dispose();` at line 70?

merks commented:
I hate to be a party pooper here  but I think a call to dispose on UI part on a non UI thread is wrong and the caller should not do that.  And this doesn't happen in a normal IDE when I exit.  But we're changing code here that will affect the normal behavior.  Normally this happens before org.eclipse.e4.core.internal.contexts.osgi.EclipseContextOSGi.dispose() is called:

<img width="519" height="357" alt="image" src="https://github.com/user-attachments/assets/c63c17e0-783e-4832-8042-7d53821f8d16" />

I really don't think we should add defensive code in the UI to deal with a situation that arises only in a test.

I highly recommend finding a way to fix this by modifying the test to dispose the UI (close the workbench) before this destroy handling kicks in.
